librz/core/golang.c

```diff
@@ -0,0 +1,964 @@
+// SPDX-FileCopyrightText: 2022 deroad <wargio@libero.it>
+// SPDX-License-Identifier: LGPL-3.0-only
+#include <rz_core.h>
+
+#define GO_MAX_STRING_SIZE 0x4000
+
+#define GO_1_2  (12)
+#define GO_1_16 (116)
+#define GO_1_18 (118)
+
+#define IS_GOPCLNTAB_1_2_LE(x)  (x[0] == 0xfb && x[1] == 0xff && x[2] == 0xff && x[3] == 0xff)
+#define IS_GOPCLNTAB_1_2_BE(x)  (x[3] == 0xfb && x[2] == 0xff && x[1] == 0xff && x[0] == 0xff)
+#define IS_GOPCLNTAB_1_16_LE(x) (x[0] == 0xfa && x[1] == 0xff && x[2] == 0xff && x[3] == 0xff)
+#define IS_GOPCLNTAB_1_16_BE(x) (x[3] == 0xfa && x[2] == 0xff && x[1] == 0xff && x[0] == 0xff)
+#define IS_GOPCLNTAB_1_18_LE(x) (x[0] == 0xf0 && x[1] == 0xff && x[2] == 0xff && x[3] == 0xff)
+#define IS_GOPCLNTAB_1_18_BE(x) (x[3] == 0xf0 && x[2] == 0xff && x[1] == 0xff && x[0] == 0xff)
+
+typedef struct go_pc_line_table_t {
+	RzIO *io;
+	ut64 vaddr;
+	ut32 size;
+	ut16 version;
+	bool big_endian;
+	ut64 text_start;
+	// quantum is the min instruction size for the program counter.
+	// i386: 1, amd64: 1, wasm: 1, s390x: 2, arm: 4, arm64: 4, mips: 4, mips: 4, ppc: 4, riscv: 4
+	ut32 quantum;
+	ut32 ptrsize;
+	ut32 nfunctab;
+	ut32 nfiletab;
+	ut32 functabsize;
+	// data offsets
+	ut64 funcnametab;
+	ut64 cutab;
+	ut64 filetab;
+	ut64 pctab;
+	ut64 funcdata;
+	ut64 functab;
+} GoPcLnTab;
+
+typedef struct go_string_recover_t {
+	RzCore *core;
+	ut64 pc;
+	ut8 *bytes;
+	ut32 size;
+	ut32 n_recovered;
+} GoStrRecover;
+
+typedef ut32 (*GoStrRecoverCb)(GoStrRecover *ctx);
+
+ut32 go_func_tab_field_size(GoPcLnTab *pclntab) {
+	if (pclntab->version >= GO_1_18) {
+		return 4;
+	}
+	return pclntab->ptrsize;
+}
+
+ut64 go_uintptr(GoPcLnTab *pclntab, ut8 *bytes) {
+	if (pclntab->ptrsize == 4) {
+		return rz_read_ble32(bytes, pclntab->big_endian);
+	}
+	return rz_read_ble64(bytes, pclntab->big_endian);
+}
+
+ut64 go_offset(GoPcLnTab *pclntab, ut32 n_word) {
+	ut8 bytes[8];
+	ut64 location = pclntab->vaddr + 8 + (n_word * pclntab->ptrsize);
+	if (0 > rz_io_nread_at(pclntab->io, location, bytes, sizeof(bytes))) {
+		return UT64_MAX;
+	}
+	return go_uintptr(pclntab, bytes);
+}
+
+ut64 go_data(GoPcLnTab *pclntab, ut32 n_word) {
+	ut64 offset = go_offset(pclntab, n_word);
+	if (offset == UT64_MAX) {
+		return UT64_MAX;
+	}
+	return pclntab->vaddr + offset;
+}
+
+#define is_addr_outside(x) ((x) <= begin || (x) >= end)
+static bool is_pclntab_valid(GoPcLnTab *pclntab) {
+	ut64 begin = pclntab->vaddr + 8;
+	ut64 end = pclntab->vaddr + pclntab->size;
+
+	if (pclntab->version > GO_1_2) {
+		if (is_addr_outside(pclntab->funcnametab)) {
+			rz_warn_if_reached();
```

Comment:
I am not sure we need warnings here, since any input can be essentially random, and it's not really a programmer's error to reach this place.

Reply:
no, is a debug feature i have introduced. let me remove them because i forgot them.